dht.go

```diff
@@ -275,7 +276,49 @@ func (dht *IpfsDHT) putLocal(key string, rec *recpb.Record) error {
 // on the given peer.
 func (dht *IpfsDHT) Update(ctx context.Context, p peer.ID) {
 	logger.Event(ctx, "updatePeer", p)
-	dht.routingTable.Update(p)
+	if dht.shouldAddPeerToRoutingTable(p) {
+		dht.routingTable.Update(p)
+	}
+}
+
+func (dht *IpfsDHT) shouldAddPeerToRoutingTable(p peer.ID) bool {
+	return dht.hasOutboundConnToPeer(p) ||
```

Comment:
Based on new discoveries, I think we should ignore if:

1. we don't have addresses for the peer; after merging https://github.com/libp2p/go-libp2p/pull/657, peers that are not routable (e.g. no NAT mapping or pinhole) nor connected to a relay will advertise 0 addrs for themselves.
2. if we only have localhost addresses for the peer (old peers without the above patch).
3. if remote multiaddr != any of peer's advertised addrs (likely to be an inbound connection via an ephemeral NAT port mapping); this correctly covers the LAN connection case, i.e. if we're on LAN, we will be connected to their advertised address, so the predicate will be satisfied.
5. if the remote advertises relay addresses (consequence of being behind a NAT).
6. if they don't support the DHT server protocol (upgraded peers).